data/collector.go

```diff
@@ -57,43 +58,53 @@ type Collector interface {
 }
 
 type collector struct {
-	clock            clock.Clock
-	captureResults   chan *v1.SensorData
-	captureErrors    chan error
-	interval         time.Duration
-	params           map[string]*anypb.Any
+	clock          clock.Clock
+	captureResults chan *v1.SensorData
+	captureErrors  chan error
+	interval       time.Duration
+	params         map[string]*anypb.Any
+	// `lock` serializes calls to `Flush` and `Close`.
 	lock             sync.Mutex
 	logger           logging.Logger
 	captureWorkers   sync.WaitGroup
 	logRoutine       sync.WaitGroup
 	cancelCtx        context.Context
 	cancel           context.CancelFunc
 	captureFunc      CaptureFunc
-	closed           bool
+	closeStarted     atomic.Bool
+	closeFinished    bool
 	target           datacapture.BufferedWriter
 	lastLoggedErrors map[string]int64
 }
 
 // Close closes the channels backing the Collector. It should always be called before disposing of a Collector to avoid
 // leaking goroutines.
 func (c *collector) Close() {
-	if c.closed {
-		return
-	}
+	// Signal the start of closing. Specifically, this is used to signal to the error logging
```

Comment:
These comments throughout are immensely helpful and boost readability by 100x, thank you.